File: server-go/notify.go
package main

import "sync"

// Notifier is the Go analogue of libqueen's pop backoff-tracker wake path
// (update_pop_backoff_tracker): a parked long-poll pop registers interest in a
// queue key and is woken the instant a push commits to that queue — instead of
// blindly re-polling. Broadcast is done by closing (and replacing) a per-key
// channel, so all current waiters wake at once.
//
// Lost-wakeup safety: a waiter MUST call Register(key) to obtain its wait
// channel BEFORE performing its pop attempt. If a push closes the channel
// between the attempt and the select, the waiter observes it already closed and
// re-attempts immediately.
type Notifier struct {
	mu    sync.Mutex
	chans map[string]chan struct{}
}

func NewNotifier() *Notifier {
	return &Notifier{chans: make(map[string]chan struct{})}
}

// Register returns the current wait channel for key, creating it if needed.
func (n *Notifier) Register(key string) <-chan struct{} {
	n.mu.Lock()
	ch, ok := n.chans[key]
	if !ok {
		ch = make(chan struct{})
		n.chans[key] = ch
	}
	n.mu.Unlock()
	return ch
}

// Notify wakes all current waiters on key (closes + drops the channel; the next
// Register creates a fresh one). Cheap no-op when nobody is parked on key.
func (n *Notifier) Notify(key string) {
	n.mu.Lock()
	if ch, ok := n.chans[key]; ok {
		close(ch)
		delete(n.chans, key)
	}
	n.mu.Unlock()
}

File: server-rust/src/notify.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tokio::sync::Notify;

/// Per-queue wake for long-poll pops: a parked empty pop waits on the queue's
/// Notify; a push to that queue wakes it (analogue of libqueen's
/// update_pop_backoff_tracker wake path). notify_waiters() only wakes tasks
/// already waiting, so callers must register (notified()) BEFORE their pop
/// attempt to avoid lost wakeups.
#[derive(Clone, Default)]
pub struct Notifier {
    map: Arc<Mutex<HashMap<String, Arc<Notify>>>>,
}

impl Notifier {
    pub fn new() -> Notifier {
        Notifier { map: Arc::new(Mutex::new(HashMap::new())) }
    }

    pub fn get(&self, key: &str) -> Arc<Notify> {
        let mut m = self.map.lock().unwrap();
        if let Some(n) = m.get(key) {
            return n.clone();
        }
        let n = Arc::new(Notify::new());
        m.insert(key.to_string(), n.clone());
        n
    }

    pub fn notify(&self, key: &str) {
        let n = {
            let m = self.map.lock().unwrap();
            m.get(key).cloned()
        };
        if let Some(n) = n {
            n.notify_waiters();
        }
    }
}
